Foreach Binary Test Refactor

crcrpar opened the pull request:
Related: https://github.com/pytorch/pytorch/issues/58833

## Changes I'm a bit concerned
- binary ops with one tensorlist and one scalarlist support complex dtypes. To realize this, I added a specialization of [`TensorListScalarListMetadata<c10::complex<double>, 1>` ](https://github.com/pytorch/pytorch/pull/59907/files#diff-131eb9b310905b15b3528da6a23e542a3a3aa952bc88f7423c98a23a8a28cca1R49). This might be out of the scope of this pull request.


cc @ptrblck @ngimel @mcarilli

crcrpar committed:
Foreach Binary Test Refactor

crcrpar commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2323,9 +2323,11 @@ def __init__(self,
-    tensors = [make_tensor((N - i, N - i), device, dtype, noncontiguous=noncontiguous) for i in range(N)]
-    return tensors
+def sample_inputs_foreach(self, device, dtype, N, *, noncontiguous=False, same_size=False):
+    if same_size:
+        return [make_tensor((N, N), device, dtype, noncontiguous=noncontiguous) for _ in range(N)]
```
This option is used by [`test_binary_op_tensors_on_different_devices`](https://github.com/pytorch/pytorch/pull/59907/files#diff-d11ea193df3e88b33bbb36334bb325f36630ad7c2486057068c58f633a79d1c1R829)

ngimel commented on test/test_foreach.py:
```diff
@@ -77,23 +139,314 @@ def _get_test_data(self, device, dtype, N):
+        op, ref, inplace_op, inplace_ref = self._get_funcs(
+            opinfo, dtype=dtype, scalars=scalar, n_tensors=N, is_fastpath=is_fastpath)
+        inputs = [opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath), scalar]
+        inputs2 = [scalar for _ in range(N)]
```
wait, this is not `scalar`, this is scalar list, you op will be `torch._for_each_add([x,x], [1,1])`, whereas what you need to test is `torch._for_each_add([x,x], 1)`. As written, this test is not at all different from `_test_binary_op_scalarlist` below.

ngimel commented on test/test_foreach.py:
```diff
@@ -57,14 +53,80 @@ def is_cuda(self):
+        return self._create_funcs(op, n_expected_cudaLaunchKernels)
+
+    # note(mkozuki): Foreach binary ops upcast 16-bits inputs to 32-bits and downcast outputs to original dtype.
+    def _requires_cast(self, op, dtype, is_fastpath, inputs, inputs2=None):
```
On second thought, this `inputs2` thing is extremely confusing. `inputs` should always be a list with two elements, inputs[1] can be a tensor list, a list of scalars or a single scalar, and there can be a code to handle that, but having both `inputs` and `inputs2` seems like an unnecessary complication

ngimel reviewed (commented):
First part of review. 
1) Please unify the tests for scalar dtypes, fast and slow path, it's hard to have a good picture of what's tested otherwise. If you want to have separate tests for this, the correct thing to do is to have decorators (but here it may be more trouble than it's worth)
2) Please in a separate PR fix the upcasting issues for addition and subtraction, we should not have special handling in tests, operations have to produce same results. It's ok that regular addition will produce slightly different results than before
3) Please streamline input/input2 handling, I don't see why input2 is needed.

crcrpar commented on test/test_foreach.py:
```diff
@@ -77,23 +139,314 @@ def _get_test_data(self, device, dtype, N):
+            return
+        if is_fastpath:
+            if requires_cast:
+                ref_inputs = [[t.to(torch.float32) for t in ref_inputs[0]], ref_inputs[1]]
```
> Also, is the difference between cast and no-cast version really big enough to fail the test? that's surprising, for alpha=1 it should be non-existent.

Without casting, the gap is (almost) proportional to the abs(alpha) if alpha is not representable with 16bits.

crcrpar commented:
Thank you for the review. I'm removing `input2` thing and unifying different scalar/scalarlist-s tests into one. So there'll be `test_binary_op_tensorlists_(fast|slow)path`, `test_binary_op_scalar_(fast|slow)path`, and `test_binary_op_scalarlist_(fast|slow)path`.

crcrpar committed:
comment for 80

crcrpar committed:
let caller calc `n_expected_cudaLaunchKernels`

also, unify
- int/float/complex/bool scalar tests -> scalar test
- int/float/complex/bool scalarlist tests -> scalarlist test

crcrpar committed:
remove inputs2 things

crcrpar committed:
update _get_funcs: pass `n_expected_cudaLaunchKernels`

ngimel commented:
Cool, thanks, do you plan on fixing regular addition so that we don't have to have `requires_cast` things?

crcrpar commented:
> Cool, thanks, do you plan on fixing regular addition so that we don't have to have `requires_cast` things?

Yes, and I wrote https://github.com/crcrpar/pytorch/commit/4627e30b5ac935b4f4e77e2d72a8410101858b48 (using accumulate type and having requires_cast retun always False) on top of this branch and local run of test_foreach.py was successful

ngimel commented:
That's cool, so can you submit PR for add separately, so that this can become simpler without the need to handle requires_cast?

ngimel commented on test/test_foreach.py:
```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+            self._regular_binary_test(dtype, op, ref, inputs, is_fastpath, alpha=alpha)
+            self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath, alpha=alpha)
+
+    @skipMeta
```
Should this be skipped for ROCm too?

crcrpar replied:
Yes. it's failing because
- this test compares the exception if raised
- foreach with meta tensors raises
- but regular with meta tensors not raise

```
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_add_meta_float32 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_add_meta_float64 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_div_meta_float32 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_div_meta_float64 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_mul_meta_float32 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_mul_meta_float64 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_sub_meta_float32 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_sub_meta_float64 - AssertionError: NotImplementedError not raised
```

ngimel commented on test/test_foreach.py:
```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+            ]):
+                disable_fastpath = op.ref == torch.div and dtype in torch.testing.get_all_int_dtypes() + [torch.bool]
+                # int scalar
+                if i == 0:
```
`if is_instance(scalar, float):` and similar things, no `enumerate`

ngimel commented on test/test_foreach.py:
```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+    @skipMeta
+    @ops(foreach_binary_op_db)
+    def test_binary_op_scalar_fastpath(self, device, dtype, op):
+        for N in N_values:
```
You can have `for N, scalar in product(N_values, scalars)`:
(you can create `scalars` tuple in advance)

ngimel commented on test/test_foreach.py:
```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+                # complex scalar
+                if i == 3:
+                    disable_fastpath |= dtype not in torch.testing.get_all_complex_dtypes()
+                print(scalar, disable_fastpath)
```
stray print

ngimel commented on test/test_foreach.py:
```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+                print(scalar, disable_fastpath)
+                self._test_binary_op_scalar(device, dtype, op, N, scalar, True, disable_fastpath)
+
+    @dtypes(*torch.testing.get_all_dtypes())
```
It's probably fine for these ops, but generally ops probably don't support `get_all_dtypes()`, so it's dangerous decorator to use

ngimel commented on test/test_foreach.py:
```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+    def test_binary_op_scalar_slowpath(self, device, dtype, op):
+        scalar_values = [
+            random.randint(1, 10), 1.0 - random.random(), True, complex(1.0 - random.random(), 1.0 - random.random())]
+        for N in N_values:
```
use product here also

ngimel commented on test/test_foreach.py:
```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+    @dtypes(*torch.testing.get_all_dtypes())
+    @ops(foreach_binary_op_db)
+    def test_binary_op_scalar_slowpath(self, device, dtype, op):
+        scalar_values = [
```
use tuple instead of list

ngimel commented on test/test_foreach.py:
```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+    def test_binary_op_scalarlist_fastpath(self, device, dtype, op):
+        for N in N_values:
+            for i, scalarlist in enumerate([
+                [random.randint(0, 9) + 1 for _ in range(N)],
```
instead of `enumerate` you can have a nested tuple there, with first element denoting type ("float", "int" or "mixed")

ngimel commented on test/test_foreach.py:
```diff
@@ -740,164 +513,133 @@ def test_add_scalar_with_empty_list_and_empty_tensor(self, device, dtype):
-            if dtype in [torch.bfloat16, torch.bool, torch.float16]:
-                tensors2 = [torch.zeros(N, N, device=device, dtype=dtype).add(2) for _ in range(N)]
+        # different devices
+        if torch.cuda.is_available() and torch.cuda.device_count() > 1:
```
Please check that `device` argument you got is `cuda`, `torch.device(device).type == "cuda"`

ngimel commented on test/test_foreach.py:
```diff
@@ -740,164 +513,133 @@ def test_add_scalar_with_empty_list_and_empty_tensor(self, device, dtype):
-        res = torch._foreach_add([tensor1], [tensor2])
-        torch._foreach_add_([tensor1], [tensor2])
-        self.assertEqual(res, [tensor1])
+        tensor1 = torch.randn(10, 10, device=device).to(dtype)
```
use `make_tensor` here

ngimel commented on test/test_foreach.py:
```diff
@@ -740,164 +513,133 @@ def test_add_scalar_with_empty_list_and_empty_tensor(self, device, dtype):
         # non contiguous
-        tensor1 = torch.randn(5, 2, 1, 3, device=device)[:, 0]
-        tensor2 = torch.randn(5, 2, 1, 3, device=device)[:, 0]
+        tensor1 = torch.randn(5, 2, 1, 3, device=device).to(dtype)[:, 0]
```
`make_tensor(noncontiguous=True)`

ngimel commented on test/test_foreach.py:
```diff
@@ -927,32 +669,34 @@ def test_unary_op_tensors_on_different_devices(self, device, dtype, op):
+    @dtypes(*torch.testing.get_all_dtypes())
+    @ops(foreach_binary_op_db)
+    def test_binary_op_tensors_on_different_devices(self, device, dtype, op):
         if self.device_type != 'cuda':
```
Use @onlyCUDA decorator

ngimel commented on test/test_foreach.py:
```diff
@@ -927,32 +669,34 @@ def test_unary_op_tensors_on_different_devices(self, device, dtype, op):
 
-    @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=True))
-    def test_binary_op_tensors_on_different_devices(self, device, dtype):
+    @dtypes(*torch.testing.get_all_dtypes())
```
you probably don't need to run this test for all the dtypes, if the goal is just to error out for the different devices

crcrpar replied:
sounds good and unary version doesn't use this decorator, thanks.

ngimel reviewed (commented):
This looks much better, thanks! @mruberry and I looked over it and left a few comments, let us know what you think. 
We could also merge fix to regular add to avoid conversions.

crcrpar commented on test/test_foreach.py:
```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+                disable_fastpath = True
+            self._test_binary_op_tensorlists(device, dtype, op, N, True, disable_fastpath)
+
+    @dtypes(*torch.testing.get_all_dtypes())
```
ah, now that foreach binary ops used `AT_DISPATCH_ALL_TYPES_AND_COMPLEX*` dispatcher, the dtypes might be the same. double check

crcrpar replied:
supported dtypes look same because `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3` is used and foreach doesn't support sparse tensors at this moment.

ngimel replied:
Yeah but that decorator doesn't make you test sparse inputs, so testing for just foreach supported dtypes (default) should be good.

crcrpar committed:
use numbers.Number and itertools.Product

crcrpar committed:
use onlyCUDA and make_tensor

crcrpar committed:
use complex alpha for complex inputs

crcrpar committed:
check device type, not cuda availability

crcrpar commented on test/test_foreach.py:
```diff
@@ -1,23 +1,29 @@
         self.func = func
 
     def __call__(self, inputs, **kwargs):
+        if len(inputs) == 2 and isinstance(inputs[1], (int, float, complex, bool)):
```
didn't know that, thanks

crcrpar committed:
unify `_regular_binary_test` and `_inplace_binary_test`

also, remove @dtypes decorator from slowpath tests

crcrpar committed:
tensorlists fastpath: skip if rocm

crcrpar committed:
remove type cast for binary ops

crcrpar committed:
line length

ngimel commented on test/test_foreach.py:
```diff
@@ -740,173 +464,141 @@ def test_add_scalar_with_empty_list_and_empty_tensor(self, device, dtype):
-        res = torch._foreach_add([tensor1], [tensor2.t()])
-        torch._foreach_add_([tensor1], [tensor2])
-        self.assertEqual(res, [tensor1])
+        inputs = ([tensor1], [tensor2])
```
should it be tensor2.t()?

crcrpar replied:
oh, thank you very much

ngimel reviewed (commented):
This looks good, only one minor comment and we can land

crcrpar committed:
transpose rhs operand tensors

crcrpar committed:
enabling kwarg dtype
